dapp/src/components/dialogs/EditMetadata.tsx

```diff
@@ -0,0 +1,234 @@
+// Copyright (c) 2025 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+'use client';
+
+import {
+    Button,
+    ButtonType,
+    Chip,
+    ChipType,
+    Dialog,
+    DialogBody,
+    DialogContent,
+    DialogPosition,
+    Header,
+    Input,
+    InputType,
+    LoadingIndicator,
+} from '@iota/apps-ui-kit';
+import { useCurrentAccount, useIotaClient, useSignAndExecuteTransaction } from '@iota/dapp-kit';
+import { ALLOWED_METADATA, isSubname, normalizeIotaName } from '@iota/iota-names-sdk';
+import { useMutation, useQueryClient } from '@tanstack/react-query';
+import { useState } from 'react';
+import toast from 'react-hot-toast';
+
+import {
+    NameRecordData,
+    NameUpdate,
+    queryKey,
+    useNameRecord,
+    useRegistrationNfts,
+    useUpdateNameTransaction,
+} from '@/hooks';
+import { getNameObject } from '@/lib/utils/names';
+
+interface EditMetadataDialogProps {
+    name: string;
+    setOpen: (bool: boolean) => void;
+}
+
+const METADATA_KEYS = [
+    { label: 'Twitter/X', allowedKey: 'twitterX' },
+    { label: 'Discord', allowedKey: 'discord' },
+    { label: 'Github', allowedKey: 'github' },
+    { label: 'Email', allowedKey: 'email' },
+    { label: 'Btc', allowedKey: 'btc' },
+    { label: 'Eth', allowedKey: 'eth' },
+    { label: 'Ltc', allowedKey: 'ltc' },
+    { label: 'Doge', allowedKey: 'doge' },
+    { label: 'Sol', allowedKey: 'sol' },
+    { label: 'Sui', allowedKey: 'sui' },
+    { label: 'Website', allowedKey: 'website' },
+    { label: 'Ipfs', allowedKey: 'ipfs' },
+    { label: 'Arweave', allowedKey: 'arweave' },
+] as const satisfies { label: string; allowedKey: keyof typeof ALLOWED_METADATA }[];
+
+const METADATA_FIELDS = METADATA_KEYS.map(({ label, allowedKey }) => ({
+    key: ALLOWED_METADATA[allowedKey],
+    label,
+    allowedKey,
+}));
+
+export function EditMetadataDialog({ name, setOpen }: EditMetadataDialogProps) {
+    const account = useCurrentAccount();
+    const iotaClient = useIotaClient();
+    const queryClient = useQueryClient();
+
+    const { data: nameRecordData } = useNameRecord(name);
+    const { data: subnamesOwned } = useRegistrationNfts('subname');
+
+    const nameRecord = nameRecordData as
+        | Extract<NameRecordData, { type: 'unavailable' }>
+        | undefined;
+
+    const initialMetadata = (() => {
+        const initial: Record<string, { selected: boolean; data: string }> = {};
+        METADATA_FIELDS.forEach(({ key }) => {
+            const data = nameRecord?.nameRecord.data[key];
+            initial[key] = {
+                selected: data !== undefined,
+                data: data || '',
+            };
+        });
+        return initial;
+    })();
+
+    const [metadata, setMetadata] = useState(initialMetadata);
```

Comment:
The `useState` accepts `T` or `() => T`, you are now recreating `T` on every render. Its wasteful

You can do:

```suggestion
    const initialMetadata = (() => {
        const initial: Record<string, { selected: boolean; data: string }> = {};
        METADATA_FIELDS.forEach(({ key }) => {
            const data = nameRecord?.nameRecord.data[key];
            initial[key] = {
                selected: data !== undefined,
                data: data || '',
            };
        });
        return initial;
    });

    const [metadata, setMetadata] = useState(initialMetadata);
```

Or (better imo):


```suggestion
    const [metadata, setMetadata] = useState(() => {
        const initial: Record<string, { selected: boolean; data: string }> = {};
        METADATA_FIELDS.forEach(({ key }) => {
            const data = nameRecord?.nameRecord.data[key];
            initial[key] = {
                selected: data !== undefined,
                data: data || '',
            };
        });
        return initial;
    });
```